From fdae024b5e21b3ff1efff56a5b201a1617bf6e25 Mon Sep 17 00:00:00 2001
From: OpenEmbedded <mario.schuknecht@dresearch-fe.de>
Date: Fri, 7 Oct 2022 18:50:04 +0200
Subject: [PATCH] imx8-isi-cap: Fix interaction with adv7280-m HYP-27448

Enable MIPI CSI channel first and the die subdevice. Otherwise the setup
with adv7280-m as subdevice does not work.
Set top/bottom in v4l2 buffer field.
Add g_std, s_std, querystd and s_input api functions. This allows to
select the input.
Set tvnorms to adv7280-m norms. This is needed for the api functions.

Signed-off-by: OpenEmbedded <mario.schuknecht@dresearch-fe.de>
---
 drivers/staging/media/imx/imx8-isi-cap.c | 235 +++++++++++++++++++++++
 1 file changed, 235 insertions(+)

diff --git a/drivers/staging/media/imx/imx8-isi-cap.c b/drivers/staging/media/imx/imx8-isi-cap.c
index 880f0023f202..8caea3f9f3e9 100644
--- a/drivers/staging/media/imx/imx8-isi-cap.c
+++ b/drivers/staging/media/imx/imx8-isi-cap.c
@@ -184,6 +184,13 @@ static int mxc_isi_update_buf_paddr(struct mxc_isi_buffer *buf, int memplanes)
 	return 0;
 }
 
+static int s_even_field = 0;
+void imx8_isi_cap_set_even_field(int even_field)
+{
+	s_even_field = even_field;
+}
+EXPORT_SYMBOL_GPL(imx8_isi_cap_set_even_field);
+
 void mxc_isi_cap_frame_write_done(struct mxc_isi_dev *mxc_isi)
 {
 	struct mxc_isi_cap_dev *isi_cap = mxc_isi->isi_cap;
@@ -214,6 +221,11 @@ void mxc_isi_cap_frame_write_done(struct mxc_isi_dev *mxc_isi)
 		vb2 = &buf->v4l2_buf.vb2_buf;
 		list_del_init(&buf->list);
 		buf->v4l2_buf.vb2_buf.timestamp = ktime_get_ns();
+		if (s_even_field) {
+			buf->v4l2_buf.field = V4L2_FIELD_BOTTOM;
+		} else {
+			buf->v4l2_buf.field = V4L2_FIELD_TOP;
+		}
 		vb2_buffer_done(&buf->v4l2_buf.vb2_buf, VB2_BUF_STATE_DONE);
 	}
 
@@ -1105,12 +1117,16 @@ static int mxc_isi_cap_streamon(struct file *file, void *priv,
 	dev_dbg(dev, "%s\n", __func__);
 
 	if (!isi_cap->is_streaming[isi_cap->id]) {
+#if 0
+		/* Enable the MIPI CSI channel first and then the subdevice. The other
+		 * way round the setup with the adv7280-m as subdevice does not work. */
 		src_sd = mxc_get_remote_subdev(&isi_cap->sd, __func__);
 		ret = (!src_sd) ? -EINVAL : v4l2_subdev_call(src_sd, core, s_power, 1);
 		if (ret) {
 			v4l2_err(&isi_cap->sd, "Call subdev s_power fail!\n");
 			return ret;
 		}
+#endif
 
 		ret = mxc_isi_config_parm(isi_cap);
 		if (ret < 0)
@@ -1132,6 +1148,14 @@ static int mxc_isi_cap_streamon(struct file *file, void *priv,
 		if (ret < 0 && ret != -ENOIOCTLCMD)
 			goto disable;
 
+#if 1
+		src_sd = mxc_get_remote_subdev(&isi_cap->sd, __func__);
+		ret = (!src_sd) ? -EINVAL : v4l2_subdev_call(src_sd, core, s_power, 1);
+		if (ret) {
+			v4l2_err(&isi_cap->sd, "Call subdev s_power fail!\n");
+			return ret;
+		}
+#endif
 		isi_cap->is_streaming[isi_cap->id] = 1;
 		mxc_isi->is_streaming = 1;
 	}
@@ -1354,6 +1378,210 @@ static int mxc_isi_cap_enum_frameintervals(struct file *file, void *fh,
 	return 0;
 }
 
+static int mxc_isi_vidioc_g_std(struct file *file, void *priv, v4l2_std_id *a)
+{
+	struct mxc_isi_cap_dev *isi_cap = video_drvdata(file);
+	struct device *dev = &isi_cap->pdev->dev;
+	struct media_entity *entity = &isi_cap->vdev.entity;
+	struct media_device *mdev = entity->graph_obj.mdev;
+	struct media_graph graph;
+	struct v4l2_subdev *subdev;
+	int ret = 0;
+
+	mutex_lock(&mdev->graph_mutex);
+
+	ret = media_graph_walk_init(&graph, entity->graph_obj.mdev);
+	if (ret) {
+		mutex_unlock(&mdev->graph_mutex);
+		return ret;
+	}
+	media_graph_walk_start(&graph, entity);
+
+	while ((entity = media_graph_walk_next(&graph))) {
+		if (!entity) {
+			dev_dbg(dev, "entity is NULL\n");
+			continue;
+		}
+
+		if (!is_media_entity_v4l2_subdev(entity)) {
+			dev_dbg(dev, "%s is no v4l2 subdev\n", entity->name);
+			continue;
+		}
+
+		subdev = media_entity_to_v4l2_subdev(entity);
+		if (!subdev) {
+			dev_dbg(dev, "%s subdev is NULL\n", entity->name);
+			continue;
+		}
+
+		ret = v4l2_subdev_call(subdev, video, g_std, a);
+		if (ret < 0 && ret != -ENOIOCTLCMD) {
+			dev_err(dev, "subdev %s g_std failed\n", subdev->name);
+			break;
+		}
+		if(ret == 0) {
+			break;
+		}
+	}
+	mutex_unlock(&mdev->graph_mutex);
+	media_graph_walk_cleanup(&graph);
+
+	return ret;
+}
+
+static int mxc_isi_vidioc_s_std(struct file *file, void *priv, v4l2_std_id a)
+{
+	struct mxc_isi_cap_dev *isi_cap = video_drvdata(file);
+	struct device *dev = &isi_cap->pdev->dev;
+	struct media_entity *entity = &isi_cap->vdev.entity;
+	struct media_device *mdev = entity->graph_obj.mdev;
+	struct media_graph graph;
+	struct v4l2_subdev *subdev;
+	int ret = 0;
+
+	mutex_lock(&mdev->graph_mutex);
+
+	ret = media_graph_walk_init(&graph, entity->graph_obj.mdev);
+	if (ret) {
+		mutex_unlock(&mdev->graph_mutex);
+		return ret;
+	}
+	media_graph_walk_start(&graph, entity);
+
+	while ((entity = media_graph_walk_next(&graph))) {
+		if (!entity) {
+			dev_dbg(dev, "entity is NULL\n");
+			continue;
+		}
+
+		if (!is_media_entity_v4l2_subdev(entity)) {
+			dev_dbg(dev, "%s is no v4l2 subdev\n", entity->name);
+			continue;
+		}
+
+		subdev = media_entity_to_v4l2_subdev(entity);
+		if (!subdev) {
+			dev_dbg(dev, "%s subdev is NULL\n", entity->name);
+			continue;
+		}
+
+		ret = v4l2_subdev_call(subdev, video, s_std, a);
+		if (ret < 0 && ret != -ENOIOCTLCMD) {
+			dev_err(dev, "subdev %s s_std failed\n", subdev->name);
+			break;
+		}
+		if(ret == 0) {
+			break;
+		}
+	}
+	mutex_unlock(&mdev->graph_mutex);
+	media_graph_walk_cleanup(&graph);
+
+	return ret;
+}
+
+static int mxc_isi_vidioc_querystd(struct file *file, void *priv, v4l2_std_id *a)
+{
+	struct mxc_isi_cap_dev *isi_cap = video_drvdata(file);
+	struct device *dev = &isi_cap->pdev->dev;
+	struct media_entity *entity = &isi_cap->vdev.entity;
+	struct media_device *mdev = entity->graph_obj.mdev;
+	struct media_graph graph;
+	struct v4l2_subdev *subdev;
+	int ret = 0;
+
+	mutex_lock(&mdev->graph_mutex);
+
+	ret = media_graph_walk_init(&graph, entity->graph_obj.mdev);
+	if (ret) {
+		mutex_unlock(&mdev->graph_mutex);
+		return ret;
+	}
+	media_graph_walk_start(&graph, entity);
+
+	while ((entity = media_graph_walk_next(&graph))) {
+		if (!entity) {
+			dev_dbg(dev, "entity is NULL\n");
+			continue;
+		}
+
+		if (!is_media_entity_v4l2_subdev(entity)) {
+			dev_dbg(dev, "%s is no v4l2 subdev\n", entity->name);
+			continue;
+		}
+
+		subdev = media_entity_to_v4l2_subdev(entity);
+		if (!subdev) {
+			dev_dbg(dev, "%s subdev is NULL\n", entity->name);
+			continue;
+		}
+
+		ret = v4l2_subdev_call(subdev, video, querystd, a);
+		if (ret < 0 && ret != -ENOIOCTLCMD) {
+			dev_err(dev, "subdev %s querystd failed\n", subdev->name);
+			break;
+		}
+		if(ret == 0) {
+			break;
+		}
+	}
+	mutex_unlock(&mdev->graph_mutex);
+	media_graph_walk_cleanup(&graph);
+
+	return ret;
+}
+
+static int mxc_isi_vidioc_s_input(struct file *file, void *priv, unsigned int input)
+{
+	struct mxc_isi_cap_dev *isi_cap = video_drvdata(file);
+	struct device *dev = &isi_cap->pdev->dev;
+	struct media_entity *entity = &isi_cap->vdev.entity;
+	struct media_device *mdev = entity->graph_obj.mdev;
+	struct media_graph graph;
+	struct v4l2_subdev *subdev;
+	int ret = 0;
+
+	mutex_lock(&mdev->graph_mutex);
+
+	ret = media_graph_walk_init(&graph, entity->graph_obj.mdev);
+	if (ret) {
+		mutex_unlock(&mdev->graph_mutex);
+		return ret;
+	}
+	media_graph_walk_start(&graph, entity);
+
+	while ((entity = media_graph_walk_next(&graph))) {
+		if (!entity) {
+			dev_dbg(dev, "entity is NULL\n");
+			continue;
+		}
+
+		if (!is_media_entity_v4l2_subdev(entity)) {
+			dev_dbg(dev, "%s is no v4l2 subdev\n", entity->name);
+			continue;
+		}
+
+		subdev = media_entity_to_v4l2_subdev(entity);
+		if (!subdev) {
+			dev_dbg(dev, "%s subdev is NULL\n", entity->name);
+			continue;
+		}
+
+		ret = v4l2_subdev_call(subdev, video, s_routing, input, 0, 0);
+		if (ret < 0 && ret != -ENOIOCTLCMD) {
+			dev_err(dev, "subdev %s s_input failed\n", subdev->name);
+			break;
+		}
+		if(ret == 0) {
+			break;
+		}
+	}
+	mutex_unlock(&mdev->graph_mutex);
+	media_graph_walk_cleanup(&graph);
+
+	return ret;
+}
+
 static const struct v4l2_ioctl_ops mxc_isi_capture_ioctl_ops = {
 	.vidioc_querycap		= mxc_isi_cap_querycap,
 
@@ -1370,6 +1598,11 @@ static const struct v4l2_ioctl_ops mxc_isi_capture_ioctl_ops = {
 	.vidioc_prepare_buf		= vb2_ioctl_prepare_buf,
 	.vidioc_create_bufs		= vb2_ioctl_create_bufs,
 
+	.vidioc_g_std			= mxc_isi_vidioc_g_std,
+	.vidioc_s_std			= mxc_isi_vidioc_s_std,
+	.vidioc_querystd		= mxc_isi_vidioc_querystd,
+	.vidioc_s_input			= mxc_isi_vidioc_s_input,
+
 	.vidioc_g_parm			= mxc_isi_cap_g_parm,
 	.vidioc_s_parm			= mxc_isi_cap_s_parm,
 
@@ -1666,6 +1899,8 @@ static int mxc_isi_register_cap_device(struct mxc_isi_cap_dev *isi_cap,
 	vdev->release	= video_device_release_empty;
 	vdev->queue	= q;
 	vdev->lock	= &isi_cap->lock;
+	/* Initialize with adv7280-m standards. */
+	vdev->tvnorms	= V4L2_STD_NTSC_M | V4L2_STD_PAL;
 
 	vdev->device_caps = V4L2_CAP_STREAMING | V4L2_CAP_VIDEO_CAPTURE_MPLANE;
 	video_set_drvdata(vdev, isi_cap);
-- 
2.35.3

